### DISCUSSION TITLE
How to sync to SMB share? Permission hell...

### DISCUSSION THREAD

**Discussion by jamess60** [abraunegg/onedrive#2768] [General]
Hi 

I tried running the container as root but quickly discovered this isn't supported. 

I tried setting the UID/GID to a none root user, but this user cant write to my SMB share (unsure why given the mount path is chowned to this user, 777d and permissions are appropriate on the file server but I digress...)

I then tried using docker volumes (as root can write to the smb share), but the container throws a "A database execution error occurred: database is locked" error. 

I only desire to sync a single onedrive account - really not sure why this cant work as root?
Any tips/workarounds?

**Comment by abraunegg** [comment]
@jamess60 
> I tried running the container as root but quickly discovered this isn't supported.

Running Docker containers as root can pose several security risks. This is why it is not supported - and you should re-think running any Docker as 'root'. Some reasons why it is a bad idea, and why this is not supported:

1. **Privilege Escalation:** If an attacker gains access to a container running as root, they may exploit vulnerabilities to escape the container and gain root access to the host system. This can lead to full control over the host, compromising its security.

2. **Limited Isolation:** Docker uses Linux namespaces and cgroups to isolate containers from the host system. However, running as root inside the container reduces this isolation. Root in a container is still root, and if the container breaks out, it has unrestricted access to the host.

3. **Increased Attack Surface:** Running processes as root increases the attack surface for vulnerabilities in the application or the container environment. Even a minor bug or misconfiguration can lead to severe consequences if the process has root privileges.

4. **Compliance and Best Practices:** Many security frameworks and compliance standards require that containers do not run as root. Following best practices, such as running containers with the least privilege necessary, is essential for maintaining a secure environment.

5. **Accidental Damage:** A process running as root has the potential to make significant, unintended changes to the container or the host system. Running with lower privileges reduces the risk of accidentally damaging the system.

> I tried setting the UID/GID to a none root user, but this user cant write to my SMB share (unsure why given the mount path is chowned to this user, 777d and permissions are appropriate on the file server but I digress...)

The user inside the Docker container is *not* the local user you are trying to write the data with, and the permissions on the server providing the SMB share is potentially where the permission fault is.

> I then tried using docker volumes (as root can write to the smb share), but the container throws a "A database execution error occurred: database is locked" error.

Again - inside the container, the volume you are trying to mount as the configuration path where the DB is being stored can't be correctly accessed or is in use by another process.

That particular error message also indicates that whatever path you are using, you may have another local process already reading|writing to that path and database - thus it is locked and cant be opened.

> I only desire to sync a single onedrive account - really not sure why this cant work as root?
> Any tips/workarounds?

Rethink your security model - do not run things as 'root'

Potentially you need to step backwards here and look at how that SMB mount is being served, the permissions there, then the permissions on the mount point and how you are mounting it, what effective permissions are being given by the 'mount' command 

All of this is *outside* the support for this client, potentially ask your question on how to correctly expose a SMB share to Linux on relevant forums for the system providing the SMB share and/or the Linux distribution you are using on how to correctly mount that share.

**Reply by jamess60** [reply]
Thanks for the quick reply. 

While I appreciate the security implications of running root containers, rootless is a second class citizen in docker. No Swarm support, no MacVLAN support, no CIFS/NFS docker volumes just to name a few. 

While I would look to run rootless containers in enterprise for the obvious security benefits, Its overkill for a one user homelab and I'm willing to accept those risks, especially considering the docker host is vlanned off and has no exposed ports. 

Surely its a bit obtuse to outright prohibit root usage? As a default behavior sure, but it would be great to see an environment variable to override this.

**Reply by abraunegg** [reply]
> Surely its a bit obtuse to outright prohibit root usage? 

Within `entrypoint.sh` you can comment out this line:

![image](https://github.com/user-attachments/assets/d9bb95c7-3ea3-459e-a179-40a26ea3ba91)

With this commented out, build a new container locally and it *should* allow you to run as 'root' (not tested)

> As a default behavior sure, but it would be great to see an environment variable to override this.

Please raise a feature request

**Comment by abraunegg** [comment]
@jamess60 
FYI this commit https://github.com/abraunegg/onedrive/pull/2769/commits/4b6b3791b66cd68444eeb4cc0fdc2fb1477e5718 has been added into RC3 which, based on some limited testing allows running as 'root' 

Please test RC3 within your environment.

**Comment by abraunegg** [comment]
@jamess60 
FYI, RC3 now includes a variable 'ONEDRIVE_RUNAS_ROOT' to flag whether to run the container as 'root' or as the 'onedrive' user:

### Normal usage, running Docker as an unprivileged user
```
alex@ubuntu-server-2204-docker:~/onedrive-pr2769$ docker run -it --name onedrive-non-root -v onedrive_conf:/onedrive/conf -v "${ONEDRIVE_DATA_DIR}:/onedrive/data" -e "ONEDRIVE_UID=${ONEDRIVE_UID}" -e "ONEDRIVE_GID=${ONEDRIVE_GID}" local-onedrive-rc3:latest
# Running container as user: onedrive
# Base Args: --monitor
# Changing ownership permissions on /onedrive/data and /onedrive/conf to onedrive:onedrive
# Launching 'onedrive' as onedrive via gosu
Reading configuration file: /onedrive/conf/config
Configuration file successfully loaded
Using IPv4 and IPv6 (if configured) for all network operations
Attempting to contact Microsoft OneDrive Login Service
Successfully reached Microsoft OneDrive Login Service
Configuring Global Azure AD Endpoints
OneDrive synchronisation interval (seconds): 300
Initialising filesystem inotify monitoring ...
Performing initial synchronisation to ensure consistent local state ...
Attempting to contact Microsoft OneDrive Login Service
Successfully reached Microsoft OneDrive Login Service
Starting a sync with Microsoft OneDrive
Fetching items from the OneDrive API for Drive ID: 66d53be8a5056eca .. 
No changes or items that can be applied were discovered while processing the data received from Microsoft OneDrive
Syncing this OneDrive Personal Shared Folder: OneDrive - Constanze
Fetching items from the OneDrive API for Drive ID: bc7d88ec1f539dcf . 
No changes or items that can be applied were discovered while processing the data received from Microsoft OneDrive
Performing a database consistency and integrity check on locally stored data .. 
Scanning the local file system '/onedrive/data' for new data to upload .. 
Performing a last examination of the most recent online data within Microsoft OneDrive to complete the reconciliation process
Fetching items from the OneDrive API for Drive ID: 66d53be8a5056eca . 
No changes or items that can be applied were discovered while processing the data received from Microsoft OneDrive
Syncing this OneDrive Personal Shared Folder: OneDrive - Constanze
Fetching items from the OneDrive API for Drive ID: bc7d88ec1f539dcf . 
No changes or items that can be applied were discovered while processing the data received from Microsoft OneDrive
Sync with Microsoft OneDrive is complete
Attempting to contact Microsoft OneDrive Login Service
Successfully reached Microsoft OneDrive Login Service
Starting a sync with Microsoft OneDrive
Syncing changes from Microsoft OneDrive ...
Sync with Microsoft OneDrive is complete
```

### Normal usage, running Docker as an unprivileged user, but using ONEDRIVE_RUNAS_ROOT flag
```
alex@ubuntu-server-2204-docker:~/onedrive-pr2769$ docker run -it --name onedrive-root -v onedrive_conf:/onedrive/conf -v "${ONEDRIVE_DATA_DIR}:/onedrive/data" -e "ONEDRIVE_UID=${ONEDRIVE_UID}" -e "ONEDRIVE_GID=${ONEDRIVE_GID}" -e ONEDRIVE_RUNAS_ROOT=1 -e ONEDRIVE_DEBUG=1 local-onedrive-rc3:latest
# Running container as root due to environment variable override
# Base Args: --monitor
# We are performing debug output
# Adding --verbose --verbose
# Changing ownership permissions on /onedrive/data and /onedrive/conf to root:root
# Launching 'onedrive' as root via gosu
DEBUG: Application started
DEBUG: Process ID: 7F78EC32CA00
DEBUG: User UID:   0
DEBUG: User GID:   0
DEBUG: User Name:  root
DEBUG: Application Version: onedrive v2.5.0-rc3-2-g4b6b379
DEBUG: Application Compiled With: LDC 2100
DEBUG: Passed in 'cliArgs': ["/usr/local/bin/onedrive", "--monitor", "--syncdir", "/onedrive/data"]
DEBUG: Note: --confdir and --verbose are not listed in 'cliArgs' array
DEBUG: Passed in --confdir if present: /onedrive/conf
DEBUG: Passed in --verbose count if present: 2
```

**Comment by abraunegg** [comment]
@jamess60 
Just to complete this out:

### root user, running docker as root, attempting to run the container:
```
root@ubuntu-server-2404-docker:~/onedrive-pr2769# docker run -it --name onedrive -v onedrive_conf:/onedrive/conf -v "${ONEDRIVE_DATA_DIR}:/onedrive/data" -e "ONEDRIVE_UID=${ONEDRIVE_UID}" -e "ONEDRIVE_GID=${ONEDRIVE_GID}" local-onedrive-rc3:latest
usermod: no changes
ROOT level privileges prohibited!
root@ubuntu-server-2404-docker:~/onedrive-pr2769#
```
This is the expected result here

### root user, running docker as root, attempting to run the container but using ONEDRIVE_RUNAS_ROOT flag
```
root@ubuntu-server-2404-docker:~/onedrive-pr2769# docker run -it --name onedrive -v onedrive_conf:/onedrive/conf -v "${ONEDRIVE_DATA_DIR}:/onedrive/data" -e "ONEDRIVE_UID=${ONEDRIVE_UID}" -e "ONEDRIVE_GID=${ONEDRIVE_GID}" -e ONEDRIVE_RUNAS_ROOT=1 local-onedrive-rc3:latest
usermod: no changes
# Running container as root due to environment variable override
# Base Args: --monitor
# Launching 'onedrive' as 0
Reading configuration file: /onedrive/conf/config
Configuration file successfully loaded
Application configuration file has been updated, checking if --resync needed
WARNING: no backup config file was found, unable to validate if any changes made
Using IPv4 and IPv6 (if configured) for all network operations
Attempting to contact Microsoft OneDrive Login Service
......
```

**Comment by jamess60** [comment]
Hi @abraunegg  - Thanks for the quick action!  :) 
Will test RC3 over the weekend, been flat out this week

**Comment by jamess60** [comment]
@abraunegg 

Hi - Apologies for the _**very**_ slow update from my side. 

Shortly after I started testing, I ran into an issue where my Microsoft account suffered a malicious sign in (despite single use secure password and 2FA!) - after resolving, I still had some issues with file integrity (random folders being created in different languages - a shockingly common issue according to [reddit](https://www.reddit.com/r/microsoft/comments/1607epn/a_worrisome_folder_creation_on_my_onedrive/)). 

As such, I have a few trust issues and have decided Onedrive isn't suitable for me (especially considering the sensitivity of the data im storing there). I'm currently seeking alternative cloud storage options for my offisite backup. 

Closing the issue alongside this comment. Thank you VERY much for the hard work you did put in to assist with this. 

Edit: Happy to donate you some beer money if you have a koffi page or similar

**Comment by abraunegg** [comment]
> Shortly after I started testing, I ran into an issue where my Microsoft account suffered a malicious sign in 
> ...
> As such, I have a few trust issues and have decided Onedrive isn't suitable for me

Yes I have even seen these weird document folders being created .. despite using clean room environment. I 100% suspect it is something with their online version of OneDrive that is causing this issue as both times this was seen, it was when I logged in via a browser in a clean Linux build system ... it certainly makes folks feel uncomfortable when it occurs - me included.

> Happy to donate you some beer money if you have a koffi page or similar

Feel free to sponsor via https://github.com/sponsors/abraunegg - thats the only way possible - sorry.

**Reply by jamess60** [reply]
No worries - yeah its certainly spooky. INSTANT panic response. You'd expect better from an entity Microsoft sized but here we are...

Beer money incoming :)
